File: examples/io_grpc/bazel_build_java/service/armeria/src/main/kotlin/com/geekinasuit/daggergrpc/iogrpc/example/armeria/dagger/ApplicationGraph.java
package com.geekinasuit.daggergrpc.iogrpc.example.armeria.dagger;

import com.geekinasuit.daggergrpc.api.ApplicationScope;
import com.geekinasuit.daggergrpc.api.GrpcApplication;
import com.geekinasuit.daggergrpc.api.GrpcCallScope;
import com.geekinasuit.daggergrpc.iogrpc.example.armeria.ExampleServer;
import dagger.Component;
import dagger.Provides;
import java.util.concurrent.atomic.AtomicInteger;
import javax.inject.Named;

@GrpcApplication(callScopedModules = {ApplicationGraph.CallModule.class})
@Component(modules = {ApplicationGraph.ApplicationModule.class, Grpc.Module.class})
@ApplicationScope
public interface ApplicationGraph extends Grpc.CallScopeGraph.Supplier {
  ExampleServer server();

  @Component.Builder
  interface Builder {
    ApplicationGraph build();
  }

  static ApplicationGraph.Builder builder() {
    return DaggerApplicationGraph.builder();
  }

  /** A module containing bindings/provisions of application-scoped (singleton) objects. */
  @dagger.Module
  interface ApplicationModule {
    @Provides
    @ApplicationScope
    @Named("app")
    static String someString() {
      return "blah";
    }
  }

  /** A module containing bindings/provisions of per-call-scoped objects. */
  @dagger.Module
  interface CallModule {
    AtomicInteger COUNTER = new AtomicInteger();

    @Provides
    @Named("call")
    @GrpcCallScope
    static Integer someInt() {
      return COUNTER.getAndIncrement();
    }
  }
}

File: io_grpc/example/service/armeria/src/main/kotlin/com/geekinasuit/daggergrpc/iogrpc/example/armeria/dagger/ApplicationGraph.kt
package com.geekinasuit.daggergrpc.iogrpc.example.armeria.dagger

import com.geekinasuit.daggergrpc.api.ApplicationScope
import com.geekinasuit.daggergrpc.iogrpc.example.armeria.ExampleServer
import dagger.Component

@Component(modules = [ApplicationGraphModule::class, GrpcHandlersModule::class])
@ApplicationScope
interface ApplicationGraph : GrpcCallScopeGraph.Supplier {
  fun server(): ExampleServer

  @Component.Builder
  interface Builder {
    fun build(): ApplicationGraph
  }

  companion object {
    fun builder(): Builder = DaggerApplicationGraph.builder()
  }
}
